File: android/app/src/main/java/com/bizarreelectronics/crm/util/ImageUploadPolicy.kt
package com.bizarreelectronics.crm.util

import android.content.Context
import android.net.Uri
import java.io.File

object ImageUploadPolicy {
    const val GENERAL_IMAGE_MAX_BYTES: Long = 10L * 1024L * 1024L
    const val SMALL_IMAGE_MAX_BYTES: Long = 5L * 1024L * 1024L
    const val FORMAT_ERROR: String =
        "Use JPEG, PNG, WebP, or GIF. HEIC/HEIF, TIFF, and DNG/RAW are not supported yet; convert to JPEG before uploading."

    private val supportedMimes = setOf("image/jpeg", "image/png", "image/webp", "image/gif")

    fun isSupportedMime(mime: String?): Boolean =
        mime?.trim()?.lowercase()?.let { supportedMimes.contains(it) } == true

    fun extensionForMime(mime: String?): String = when (mime?.trim()?.lowercase()) {
        "image/png" -> "png"
        "image/webp" -> "webp"
        "image/gif" -> "gif"
        else -> "jpg"
    }

    fun formatSize(bytes: Long): String =
        if (bytes % (1024L * 1024L) == 0L) "${bytes / (1024L * 1024L)} MB" else "${bytes / 1024L} KB"

    fun sizeOf(context: Context, uri: Uri): Long =
        context.contentResolver.openFileDescriptor(uri, "r")?.use { it.statSize } ?: -1L

    fun validate(context: Context, uri: Uri, maxBytes: Long): String? {
        val mime = context.contentResolver.getType(uri)
        if (!isSupportedMime(mime)) return FORMAT_ERROR
        val size = sizeOf(context, uri)
        if (size == 0L) return "Selected image is empty."
        if (size > maxBytes) return "Image exceeds the ${formatSize(maxBytes)} size limit."
        return null
    }

    fun validate(file: File, mime: String, maxBytes: Long): String? {
        if (!isSupportedMime(mime)) return FORMAT_ERROR
        if (file.length() == 0L) return "Selected image is empty."
        if (file.length() > maxBytes) return "Image exceeds the ${formatSize(maxBytes)} size limit."
        return null
    }
}

File: packages/server/src/utils/imageUploadPolicy.ts
export const IMAGE_UPLOAD_MIME_TYPES = ['image/jpeg', 'image/png', 'image/webp', 'image/gif'] as const;
export type ImageUploadMimeType = (typeof IMAGE_UPLOAD_MIME_TYPES)[number];

export const IMAGE_UPLOAD_EXTENSIONS = new Set(['.jpg', '.jpeg', '.png', '.webp', '.gif']);

export const IMAGE_UPLOAD_FORMAT_LABEL = 'JPEG, PNG, WebP, or GIF';
export const IMAGE_UPLOAD_UNSUPPORTED_ADVANCED_FORMATS = 'HEIC/HEIF, TIFF, and DNG/RAW';
export const IMAGE_UPLOAD_FORMAT_ERROR =
  `Only ${IMAGE_UPLOAD_FORMAT_LABEL} images are supported. ` +
  `${IMAGE_UPLOAD_UNSUPPORTED_ADVANCED_FORMATS} uploads need server-side conversion first; convert to JPEG before uploading.`;

export const GENERAL_IMAGE_UPLOAD_MAX_BYTES = 10 * 1024 * 1024;
export const SMALL_IMAGE_UPLOAD_MAX_BYTES = 5 * 1024 * 1024;
export const INLINE_LOGO_MAX_BYTES = 500_000;

const MIME_TO_EXTENSION: Record<ImageUploadMimeType, string> = {
  'image/jpeg': '.jpg',
  'image/png': '.png',
  'image/webp': '.webp',
  'image/gif': '.gif',
};

export function isSupportedImageMime(mime: string | undefined | null): mime is ImageUploadMimeType {
  return IMAGE_UPLOAD_MIME_TYPES.includes((mime || '').trim().toLowerCase() as ImageUploadMimeType);
}

export function imageExtensionForMime(mime: string | undefined | null): string | null {
  const normalized = (mime || '').trim().toLowerCase();
  if (!isSupportedImageMime(normalized)) return null;
  return MIME_TO_EXTENSION[normalized];
}

export function sanitizedImageExtension(originalName: string): string | null {
  const ext = originalName ? originalName.toLowerCase().match(/\.[a-z0-9]+$/)?.[0] : null;
  return ext && IMAGE_UPLOAD_EXTENSIONS.has(ext) ? ext : null;
}

export function formatUploadSize(bytes: number): string {
  if (bytes % (1024 * 1024) === 0) return `${bytes / (1024 * 1024)} MB`;
  if (bytes % 1024 === 0) return `${bytes / 1024} KB`;
  return `${bytes} bytes`;
}

export function imageUploadSizeError(maxBytes: number): string {
  return `Image exceeds the ${formatUploadSize(maxBytes)} size limit`;
}
